README.md
## WebServer_ESP32_SC_W5500 Library

[![arduino-library-badge](https://www.ardu-badge.com/badge/WebServer_ESP32_SC_W5500.svg?)](https://www.ardu-badge.com/WebServer_ESP32_SC_W5500)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/WebServer_ESP32_SC_W5500.svg)](https://github.com/khoih-prog/WebServer_ESP32_SC_W5500/releases)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/WebServer_ESP32_SC_W5500.svg)](http://github.com/khoih-prog/WebServer_ESP32_SC_W5500/issues)


<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Donate to my libraries using BuyMeACoffee" style="height: 50px !important;width: 181px !important;" ></a>
<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-orange.svg?logo=buy-me-a-coffee&logoColor=FFDD00" style="height: 20px !important;width: 200px !important;" ></a>

---
---

## Table of Contents

* [Why do we need this WebServer_ESP32_SC_W5500 library](#why-do-we-need-this-WebServer_ESP32_SC_W5500-library)
  * [Features](#features)
  * [Currently supported Boards](#currently-supported-boards)
* [Changelog](changelog.md)
* [Prerequisites](#prerequisites)
* [Installation](#installation)
  * [Use Arduino Library Manager](#use-arduino-library-manager)
  * [Manual Install](#manual-install)
  * [VS Code & PlatformIO](#vs-code--platformio)
* [Libraries' Patches](#libraries-patches)
  * [1. For fixing ESP32 compile error](#1-for-fixing-esp32-compile-error)
* [HOWTO Fix `Multiple Definitions` Linker Error](#howto-fix-multiple-definitions-linker-error)
* [HOWTO Use analogRead() with ESP32 running WiFi and/or BlueTooth (BT/BLE)](#howto-use-analogread-with-esp32-running-wifi-andor-bluetooth-btble)
  * [1. ESP32 has 2 ADCs, named ADC1 and ADC2](#1--esp32-has-2-adcs-named-adc1-and-adc2)
  * [2. ESP32 ADCs functions](#2-esp32-adcs-functions)
  * [3. ESP32 WiFi uses ADC2 for WiFi functions](#3-esp32-wifi-uses-adc2-for-wifi-functions)
* [Usage](#usage)
  * [Class Constructor](#class-constructor)
  * [Basic Operation](#basic-operations)
  * [Advanced Options](#advanced-options)
  * [Other Function Calls](#other-function-calls)
* [How to connect W5500 to ESP32_S2/S3/C3](#How-to-connect-W5500-to-ESP32_S2S3C3)
* [Examples](#examples)
  * [Original Examples](#original-examples)
    * [ 1. AdvancedWebServer](examples/AdvancedWebServer)
    * [ 2. HelloServer](examples/HelloServer)
    * [ 3. HelloServer2](examples/HelloServer2)
    * [ 4. HttpBasicAuth](examples/HttpBasicAuth)
    * [ 5. **MQTTClient_Auth**](examples/MQTTClient_Auth)
    * [ 6. **MQTTClient_Basic**](examples/MQTTClient_Basic)
    * [ 7. **MQTT_ThingStream**](examples/MQTT_ThingStream)
    * [ 8. PostServer](examples/PostServer)
    * [ 9. SimpleAuthentication](examples/SimpleAuthentication)
    * [10. UdpNTPClient](examples/UdpNTPClient)
    * [11. UdpSendReceive](examples/UdpSendReceive)
    * [12. WebClient](examples/WebClient)
    * [13. WebClientRepeating](examples/WebClientRepeating)
    * [14. WebServer](examples/WebServer)
    * [15. **multiFileProject**](examples/multiFileProject)
* [Example AdvancedWebServer](#example-advancedwebserver)
  * [File AdvancedWebServer.ino](#file-advancedwebserverino)
* [Debug Terminal Output Samples](#debug-terminal-output-samples)
  * [ 1. AdvancedWebServer on ESP32S3_DEV with ESP32_S3_W5500](#1-AdvancedWebServer-on-ESP32S3_DEV-with-ESP32_S3_W5500)
  * [ 2. MQTT_ThingStream on ESP32S3_DEV with ESP32_S3_W5500](#2-MQTT_ThingStream-on-ESP32S3_DEV-with-ESP32_S3_W5500)
  * [ 3. MQTTClient_Auth on ESP32S3_DEV with ESP32_S3_W5500](#3-MQTTClient_Auth-on-ESP32S3_DEV-with-ESP32_S3_W5500)
  * [ 4. MQTTClient_Basic on ESP32S3_DEV with ESP32_S3_W5500](#4-MQTTClient_Basic-on-ESP32S3_DEV-with-ESP32_S3_W5500)
  * [ 5. WebClient on ESP32S3_DEV with ESP32_S3_W5500](#5-WebClient-on-ESP32S3_DEV-with-ESP32_S3_W5500)
  * [ 6. UDPSendReceive on ESP32S3_DEV with ESP32_S3_W5500](#6-UDPSendReceive-on-ESP32S3_DEV-with-ESP32_S3_W5500)
  * [ 7. AdvancedWebServer on ESP32S2_DEV with ESP32_S2_W5500](#7-AdvancedWebServer-on-ESP32S2_DEV-with-ESP32_S2_W5500)
  * [ 8. AdvancedWebServer on ESP32C3_DEV with ESP32_C3_W5500](#8-AdvancedWebServer-on-ESP32C3_DEV-with-ESP32_C3_W5500)
* [Debug](#debug)
* [Troubleshooting](#troubleshooting)
* [Releases](#releases)
* [Issues](#issues)
* [TO DO](#to-do)
* [DONE](#done)
* [Contributions and Thanks](#contributions-and-thanks)
* [Contributing](#contributing)
* [License](#license)
* [Copyright](#copyright)


---
---

### Why do we need this [WebServer_ESP32_SC_W5500 library](https://github.com/khoih-prog/WebServer_ESP32_SC_W5500)

#### Features

This [**WebServer_ESP32_SC_W5500 library**](https://github.com/khoih-prog/WebServer_ESP32_SC_W5500) is a simple yet complete WebServer wrapper library for **ESP32_S2/S3/C3-based** boards using `LwIP W5500 Ethernet`. The functions are similar and compatible to those of [`ESP32 WebServer`](https://github.com/espressif/arduino-esp32/tree/master/libraries/WebServer) and [`ESP8266WebServer`](https://github.com/esp8266/Arduino/tree/master/libraries/ESP8266WebServer) libraries to make life much easier to port sketches from ESP8266/ESP32.

This [**WebServer_ESP32_SC_W5500 library**](https://github.com/khoih-prog/WebServer_ESP32_SC_W5500) also provides high-level **HTTP(S), MQTT(S) and WebSockets Client** by using the functions from ESP32's [**HTTPClient Library**](https://github.com/espressif/arduino-esp32/tree/master/libraries/HTTPClient)

The library provides these features:

1. TCP Server and Client
2. UDP Server and Client
3. HTTP(S) Server and Client
4. HTTP(S) GET and POST requests, provides argument parsing, handles one client at a time.
5. **High-level HTTP(S) (GET, POST, PUT, PATCH, DELETE), MQTT(S) and WebSocket Client**.

Library is providing similar functions to:

1. [Ivan Grokhotkov's ESP32 WebServer](https://github.com/espressif/arduino-esp32/tree/master/libraries/WebServer)
2. [HTTPClient Library](https://github.com/espressif/arduino-esp32/tree/master/libraries/HTTPClient)

---

Please also check these twin libraries

1. [WebServer_WT32_ETH01](https://github.com/khoih-prog/WebServer_WT32_ETH01) for ESP32-based `WT32_ETH01` using `LwIP LAN8720`
2. [WebServer_ESP32_ENC](https://github.com/khoih-prog/WebServer_ESP32_ENC) for ESP32-boards using `LwIP ENC28J60`
3. [WebServer_ESP32_W5500](https://github.com/khoih-prog/WebServer_ESP32_W5500) for ESP32-boards using `LwIP W5500`
4. [WebServer_ESP32_SC_ENC](https://github.com/khoih-prog/WebServer_ESP32_SC_ENC) for ESP32_S3-boards using `LwIP ENC28J60`

and more complex libraries for ESP32-based using `LwIP Ethernet`

1. [AsyncHTTPRequest_ESP32_Ethernet](https://github.com/khoih-prog/AsyncHTTPRequest_ESP32_Ethernet)
2. [AsyncHTTPSRequest_ESP32_Ethernet](https://github.com/khoih-prog/AsyncHTTPSRequest_ESP32_Ethernet)
3. [AsyncMQTT_ESP32](https://github.com/khoih-prog/AsyncMQTT_ESP32)
4. [HTTPS_Server_Generic](https://github.com/khoih-prog/HTTPS_Server_Generic)


---

#### Currently supported Boards

This [**WebServer_ESP32_SC_W5500** library](https://github.com/khoih-prog/WebServer_ESP32_SC_W5500) currently supports these following boards:

 1. **ESP32_S3 boards** using `LwIP W5500 Ethernet`
 2. **ESP32_S2 boards** using `LwIP W5500 Ethernet`
 3. **ESP32_C3 boards** using `LwIP W5500 Ethernet`
 
---

#### ESP32S2_DEV

<p align="center">
    <img src="https://github.com/khoih-prog/WebServer_ESP32_SC_W5500/raw/main/pics/ESP32S2_DEV.png">
</p> 

#### ESP32S3_DEV

<p align="center">
    <img src="https://github.com/khoih-prog/WebServer_ESP32_SC_W5500/raw/main/pics/ESP32S3_DEV.png">
</p> 

#### ESP32C3_DEV

<p align="center">
    <img src="https://github.com/khoih-prog/WebServer_ESP32_SC_W5500/raw/main/pics/ESP32_C3_DevKitC_02.png">
</p> 

---

#### W5500

<p align="center">
    <img src="https://github.com/khoih-prog/WebServer_ESP32_SC_W5500/raw/main/pics/W5500.png">
</p>
 
<p align="center">
    <img src="https://github.com/khoih-prog/WebServer_ESP32_SC_W5500/raw/main/pics/W5500_small.png">
</p> 
 
---
---


## Prerequisites

 1. [`Arduino IDE 1.8.19+` for Arduino](https://github.com/arduino/Arduino). [![GitHub release](https://img.shields.io/github/release/arduino/Arduino.svg)](https://github.com/arduino/Arduino/releases/latest)
 2. [`ESP32 Core 2.0.5+`](https://github.com/espressif/arduino-esp32) for ESP32-based ESP32S2/S3/C3_DEV with W5500 boards. [![Latest release](https://img.shields.io/github/release/espressif/arduino-esp32.svg)](https://github.com/espressif/arduino-esp32/releases/latest/)


---
---

## Installation

### Use Arduino Library Manager

The best and easiest way is to use `Arduino Library Manager`. Search for `WebServer_ESP32_SC_W5500`, then select / install the latest version.
You can also use this link [![arduino-library-badge](https://www.ardu-badge.com/badge/WebServer_ESP32_SC_W5500.svg?)](https://www.ardu-badge.com/WebServer_ESP32_SC_W5500) for more detailed instructions.

### Manual Install

1. Navigate to [WebServer_ESP32_SC_W5500](https://github.com/khoih-prog/WebServer_ESP32_SC_W5500) page.
2. Download the latest release `WebServer_ESP32_SC_W5500-main.zip`.
3. Extract the zip file to `WebServer_ESP32_SC_W5500-main` directory 
4. Copy the whole `WebServer_ESP32_SC_W5500-main` folder to Arduino libraries' directory such as `~/Arduino/libraries/`.

### VS Code & PlatformIO:

1. Install [VS Code](https://code.visualstudio.com/)
2. Install [PlatformIO](https://platformio.org/platformio-ide)
3. Install [**WebServer_ESP32_SC_W5500** library](https://registry.platformio.org/libraries/khoih-prog/WebServer_ESP32_SC_W5500) by using [Library Manager](https://registry.platformio.org/libraries/khoih-prog/WebServer_ESP32_SC_W5500/installation). Search for WebServer_ESP32_SC_W5500 in [Platform.io Author's Libraries](https://platformio.org/lib/search?query=author:%22Khoi%20Hoang%22)
4. Use included [platformio.ini](platformio/platformio.ini) file from examples to ensure that all dependent libraries will installed automatically. Please visit documentation for the other options and examples at [Project Configuration File](https://docs.platformio.org/page/projectconf.html)

---
---


### Libraries' Patches

#### 1. For fixing ESP32 compile error

To fix [`ESP32 compile error`](https://github.com/espressif/arduino-esp32), just copy the following file into the [`ESP32`](https://github.com/espressif/arduino-esp32) cores/esp32 directory (e.g. ./arduino-1.8.19/hardware/espressif/cores/esp32) to overwrite the old file:
- [Server.h](LibraryPatches/esp32/cores/esp32/Server.h)


---
---


### HOWTO Fix `Multiple Definitions` Linker Error

The current library implementation, using `xyz-Impl.h` instead of standard `xyz.cpp`, possibly creates certain `Multiple Definitions` Linker error in certain use cases.

You can include this `.hpp` file

```cpp
// Can be included as many times as necessary, without `Multiple Definitions` Linker Error
#include "WebServer_ESP32_SC_W5500.hpp"     //https://github.com/khoih-prog/WebServer_ESP32_SC_W5500
```

in many files. But be sure to use the following `.h` file **in just 1 `.h`, `.cpp` or `.ino` file**, which must **not be included in any other file**, to avoid `Multiple Definitions` Linker Error

```cpp
// To be included only in main(), .ino with setup() to avoid `Multiple Definitions` Linker Error
#include "WebServer_ESP32_SC_W5500.h"           //https://github.com/khoih-prog/WebServer_ESP32_SC_W5500
```

Check the new [**multiFileProject** example](examples/multiFileProject) for a `HOWTO` demo.

---
---


### HOWTO Use analogRead() with ESP32 running WiFi and/or BlueTooth (BT/BLE)

Please have a look at [**ESP_WiFiManager Issue 39: Not able to read analog port when using the autoconnect example**](https://github.com/khoih-prog/ESP_WiFiManager/issues/39) to have more detailed description and solution of the issue.

#### 1.  ESP32 has 2 ADCs, named ADC1 and ADC2

#### 2. ESP32 ADCs functions

- `ADC1` controls ADC function for pins **GPIO32-GPIO39**
- `ADC2` controls ADC function for pins **GPIO0, 2, 4, 12-15, 25-27**

#### 3. ESP32 WiFi uses ADC2 for WiFi functions

Look in file [**adc_common.c**](https://github.com/espressif/esp-idf/blob/master/components/driver/adc_common.c#L61)

> In ADC2, there're two locks used for different cases:
> 1. lock shared with app and Wi-Fi:
>    ESP32:
>         When Wi-Fi using the ADC2, we assume it will never stop, so app checks the lock and returns immediately if failed.
>    ESP32S2:
>         The controller's control over the ADC is determined by the arbiter. There is no need to control by lock.
> 
> 2. lock shared between tasks:
>    when several tasks sharing the ADC2, we want to guarantee
>    all the requests will be handled.
>    Since conversions are short (about 31us), app returns the lock very soon,
>    we use a spinlock to stand there waiting to do conversions one by one.
> 
> adc2_spinlock should be acquired first, then adc2_wifi_lock or rtc_spinlock.


- In order to use `ADC2` for other functions, we have to **acquire complicated firmware locks and very difficult to do**
- So, it's not advisable to use `ADC2` with WiFi/BlueTooth (BT/BLE).
- Use `ADC1`, and pins GPIO32-GPIO39
- If somehow it's a must to use those pins serviced by `ADC2` (**GPIO0, 2, 4, 12, 13, 14, 15, 25, 26 and 27**), use the **fix mentioned at the end** of [**ESP_WiFiManager Issue 39: Not able to read analog port when using the autoconnect example**](https://github.com/khoih-prog/ESP_WiFiManager/issues/39) to work with ESP32 WiFi/BlueTooth (BT/BLE).


---
---

## Usage


### Class Constructor

```cpp
WebServer server(80);
```

Creates the `WebServer_ESP32_SC_W5500` class object.

**Parameters:**
 
host port number: ``int port`` (default is the standard HTTP port 80)

---

### Basic Operations

#### Starting the server

```cpp
void begin();
```

#### Handling incoming client requests

```cpp
void handleClient();
```

#### Disabling the server

```cpp
void close();
void stop();
```

Both methods function the same

#### Client request handlers

```cpp
void on();
void addHandler();
void onNotFound();
void onFileUpload();	
```

Example:

```cpp
server.on("/", handlerFunction);
server.onNotFound(handlerFunction);   // called when handler is not assigned
server.onFileUpload(handlerFunction); // handle file uploads
```

#### Sending responses to the client

```cpp
void send();
void send_P();
```

**Parameters:**

`code` - HTTP response code, can be `200` or `404`, etc.

`content_type` - HTTP content type, like `"text/plain"` or `"image/png"`, etc.

`content` - actual content body

---

### Advanced Options

#### Getting information about request arguments

```cpp
const String & arg();
const String & argName();
int args();
bool hasArg();
```

**Function usage:**

`arg` - get request argument value, use `arg("plain")` to get POST body
	
`argName` - get request argument name
	
`args` - get arguments count
	
`hasArg` - check if argument exist


#### Getting information about request headers

```cpp
const String & header();
const String & headerName();
const String & hostHeader();
int headers();
bool hasHeader();
```

**Function usage:**

`header` - get request header value

`headerName` - get request header name

`hostHeader` - get request host header if available, else empty string
  
`headers` - get header count
	
`hasHeader` - check if header exist


#### Authentication

```cpp
bool authenticate();
void requestAuthentication();
```

**Function usage:**

`authenticate` - server authentication, returns true if client is authenticated else false

`requestAuthentication` - sends authentication failure response to the client

**Example Usage:**

```cpp
if(!server.authenticate(username, password))
{
  server.requestAuthentication();
}
```

#### Other Function Calls

```cpp
const String & uri();     // get the current uri
HTTPMethod  method();     // get the current method 
WiFiClient client();      // get the current client
HTTPUpload & upload();    // get the current upload
void setContentLength();  // set content length
void sendHeader();        // send HTTP header
void sendContent();       // send content
void sendContent_P(); 
void collectHeaders();    // set the request headers to collect
void serveStatic();
size_t streamFile();
```


---
---

### How to connect W5500 to ESP32_S2/S3/C3

You can change the `INT` pin to another one. Default is `GPIO4`

```cpp
// Must connect INT to GPIOxx or not working
#define INT_GPIO            4
```

---

#### W5500

<p align="center">
    <img src="https://github.com/khoih-prog/WebServer_ESP32_SC_W5500/raw/main/pics/W5500.png">
</p>
 
<p align="center">
    <img src="https://github.com/khoih-prog/WebServer_ESP32_SC_W5500/raw/main/pics/W5500_small.png">
</p> 
 
---


#### ESP32S3_DEV

<p align="center">
    <img src="https://github.com/khoih-prog/WebServer_ESP32_SC_W5500/raw/main/pics/ESP32S3_DEV.png">
</p> 


|W5500|<--->|ESP32_S3|
|:-:|:-:|:-:|
|MOSI|<--->|GPIO11|
|MISO|<--->|GPIO13|
|SCK|<--->|GPIO12|
|SS|<--->|GPIO10|
|INT|<--->|GPIO4|
|RST|<--->|RST|
|GND|<--->|GND|
|3.3V|<--->|3.3V|

---

#### ESP32S2_DEV

<p align="center">
    <img src="https://github.com/khoih-prog/WebServer_ESP32_SC_W5500/raw/main/pics/ESP32S2_DEV.png">
</p> 


|W5500|<--->|ESP32_S2|
|:-:|:-:|:-:|
|MOSI|<--->|GPIO35|
|MISO|<--->|GPIO37|
|SCK|<--->|GPIO36|
|SS|<--->|GPIO34|
|INT|<--->|GPIO4|
|RST|<--->|RST|
|GND|<--->|GND|
|3.3V|<--->|3.3V|

---


#### ESP32C3_DEV

<p align="center">
    <img src="https://github.com/khoih-prog/WebServer_ESP32_SC_W5500/raw/main/pics/ESP32_C3_DevKitC_02.png">
</p> 


|W5500|<--->|ESP32_C3|
|:-:|:-:|:-:|
|MOSI|<--->|GPIO6|
|MISO|<--->|GPIO5|
|SCK|<--->|GPIO4|
|SS|<--->|GPIO7|
|INT|<--->|GPIO10|
|RST|<--->|RST|
|GND|<--->|GND|
|3.3V|<--->|3.3V|


---
---

### Examples:

#### Original Examples

 1. [AdvancedWebServer](examples/AdvancedWebServer)
 2. [HelloServer](examples/HelloServer)
 3. [HelloServer2](examples/HelloServer2)
 4. [HttpBasicAuth](examples/HttpBasicAuth)
 5. [**MQTTClient_Auth**](examples/MQTTClient_Auth)
 6. [**MQTTClient_Basic**](examples/MQTTClient_Basic)
 7. [**MQTT_ThingStream**](examples/MQTT_ThingStream)
 8. [PostServer](examples/PostServer)
 9. [SimpleAuthentication](examples/SimpleAuthentication)
10. [UdpNTPClient](examples/UdpNTPClient)
11. [UdpSendReceive](examples/UdpSendReceive)
12. [WebClient](examples/WebClient)
13. [WebClientRepeating](examples/WebClientRepeating)
14. [WebServer](examples/WebServer)
15. [**multiFileProject**](examples/multiFileProject)

---
---

### Example [AdvancedWebServer](examples/AdvancedWebServer)

#### File [AdvancedWebServer.ino](examples/AdvancedWebServer/AdvancedWebServer.ino)


https://github.com/khoih-prog/WebServer_ESP32_SC_W5500/blob/de023206ab284ab9e0cfd063316ea0946eb5eff0/examples/AdvancedWebServer/AdvancedWebServer.ino#L41-L293

---
---

### Debug Terminal Output Samples

#### 1. AdvancedWebServer on ESP32S3_DEV with ESP32_S3_W5500

The following are debug terminal output and screen shot when running example [AdvancedWebServer](examples/AdvancedWebServer) on `ESP32S3_DEV` with `ESP32_S3_W5500`

<p align="center">
    <img src="https://github.com/khoih-prog/WebServer_ESP32_SC_W5500/raw/main/pics/AdvancedWebServer.png">
</p>

```cpp
Start AdvancedWebServer on ESP32S3_DEV with ESP32_S3_W5500
WebServer_ESP32_SC_W5500 v1.0.1 for core v2.0.0+
[EWS] Default SPI pinout:
[EWS] SPI_HOST: 2
[EWS] MOSI: 11
[EWS] MISO: 13
[EWS] SCK: 12
[EWS] CS: 10
[EWS] INT: 4
[EWS] SPI Clock (MHz): 25
[EWS] =========================

ETH Started
ETH Connected
ETH MAC: DE:AD:BE:EF:BE:0E, IPv4: 192.168.2.106
FULL_DUPLEX, 100Mbps
HTTP EthernetWebServer is @ IP : 192.168.2.106
.......... ..........
```

---

#### 2. MQTT_ThingStream on ESP32S3_DEV with ESP32_S3_W5500

The terminal output of **ESP32S3_DEV with W5500** running [MQTT_ThingStream example](examples/MQTT_ThingStream)


```cpp

changelog.md
## WebServer_ESP32_SC_W5500 Library

[![arduino-library-badge](https://www.ardu-badge.com/badge/WebServer_ESP32_SC_W5500.svg?)](https://www.ardu-badge.com/WebServer_ESP32_SC_W5500)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/WebServer_ESP32_SC_W5500.svg)](https://github.com/khoih-prog/WebServer_ESP32_SC_W5500/releases)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/WebServer_ESP32_SC_W5500.svg)](http://github.com/khoih-prog/WebServer_ESP32_SC_W5500/issues)

<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Donate to my libraries using BuyMeACoffee" style="height: 50px !important;width: 181px !important;" ></a>
<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-orange.svg?logo=buy-me-a-coffee&logoColor=FFDD00" style="height: 20px !important;width: 200px !important;" ></a>

---
---


## Table of Contents


* [Changelog](#changelog)
  * [Releases v1.2.0](#releases-v120)
  * [Releases v1.1.0](#releases-v110)
  * [Releases v1.0.1](#releases-v101)
  * [Releases v1.0.0](#releases-v100)

---
---

## Changelog

#### Releases v1.2.0

1. Add support to ESP32_C3-based boards using `LwIP W5500 Ethernet`

#### Releases v1.1.0

1. Add support to ESP32_S2-based boards using `LwIP W5500 Ethernet`

#### Releases v1.0.1

1. Using `SPI_DMA_CH_AUTO` instead of manually selected
2. Add example [multiFileProject](https://github.com/khoih-prog/WebServer_ESP32_SC_W5500/tree/main/examples/multiFileProject)

#### Releases v1.0.0

1. Initial coding to support ESP32_S3-based boards using `LwIP W5500 Ethernet`.
2. Use `allman astyle`




src/WebServer_ESP32_SC_W5500.hpp
/****************************************************************************************************************************
  WebServer_ESP32_SC_W5500.h
  
  For Ethernet shields using ESP32_SC_W5500 (ESP32_S2/S3/C3 + LwIP W5500)

  WebServer_ESP32_SC_W5500 is a library for the ESP32_S2/S3/C3 with LwIP Ethernet W5500

  Based on and modified from ESP8266 https://github.com/esp8266/Arduino/releases
  Built by Khoi Hoang https://github.com/khoih-prog/WebServer_ESP32_SC_W5500
  Licensed under GPLv3 license

  Version: 1.2.0

  Version Modified By   Date      Comments
  ------- -----------  ---------- -----------
  1.0.0   K Hoang      13/12/2022 Initial coding for ESP32_S3_W5500 (ESP32_S3 + W5500)
  1.0.1   K Hoang      14/12/2022 Using SPI_DMA_CH_AUTO instead of manually selected
  1.1.0   K Hoang      19/12/2022 Add support to ESP32_S2_W5500 (ESP32_S2 + W5500)
  1.2.0   K Hoang      20/12/2022 Add support to ESP32_C3_W5500 (ESP32_C3 + W5500)
 *****************************************************************************************************************************/

#pragma once

#ifndef WEBSERVER_ESP32_SC_W5500_HPP
#define WEBSERVER_ESP32_SC_W5500_HPP

/////////////////////////////////////////////////////////////

#if ( ARDUINO_ESP32S2_DEV || ARDUINO_FEATHERS2 || ARDUINO_ESP32S2_THING_PLUS || ARDUINO_MICROS2 || \
      ARDUINO_METRO_ESP32S2 || ARDUINO_MAGTAG29_ESP32S2 || ARDUINO_FUNHOUSE_ESP32S2 || \
      ARDUINO_ADAFRUIT_FEATHER_ESP32S2_NOPSRAM )
  #if (_ETHERNET_WEBSERVER_LOGLEVEL_ > 3)
		#warning Using ESP32_S2. To follow library instructions to install esp32-s2 core and WebServer Patch
		#warning You have to select HUGE APP or 1.9-2.0 MB APP to be able to run Config Portal. Must use PSRAM
  #endif
  
  #define USING_ESP32_S2        true
  
  #ifndef SHIELD_TYPE
  	#define SHIELD_TYPE         "ESP32_S2_W5500"
	#endif
   
#elif ( ARDUINO_ESP32C3_DEV )
	#if (_ETHERNET_WEBSERVER_LOGLEVEL_ > 3)
		#if ( defined(ESP_ARDUINO_VERSION_MAJOR) && (ESP_ARDUINO_VERSION_MAJOR >= 2) )
		  #warning Using ESP32_C3 using core v2.0.0+. Either LittleFS, SPIFFS or EEPROM OK.
		#else
		  #warning Using ESP32_C3 using core v1.0.6-. To follow library instructions to install esp32-c3 core. Only SPIFFS and EEPROM OK.
		#endif
		
		#warning You have to select Flash size 2MB and Minimal APP (1.3MB + 700KB) for some boards
  #endif
    
  #define USING_ESP32_C3        true
  
  #ifndef SHIELD_TYPE
  	#define SHIELD_TYPE         "ESP32_C3_W5500"
	#endif
  
  //#error ESP32_C3 not supported yet
  
#elif ( defined(ARDUINO_ESP32S3_DEV) || defined(ARDUINO_ESP32_S3_BOX) || defined(ARDUINO_TINYS3) || \
        defined(ARDUINO_PROS3) || defined(ARDUINO_FEATHERS3) )
  #if (_ETHERNET_WEBSERVER_LOGLEVEL_ > 3)
  	#warning Using ESP32_S3. To install esp32-s3-support branch if using core v2.0.2-
  #endif
  
  #define USING_ESP32_S3        true
  
  #ifndef SHIELD_TYPE
  	#define SHIELD_TYPE         "ESP32_S3_W5500"
	#endif
 
#else
	#error This code is intended to run on the ESP32_S2/3, ESP32_C3 platform! Please check your Tools->Board setting.  
#endif

#define BOARD_NAME      ARDUINO_BOARD

//////////////////////////////////////////////////////////////

//#define CONFIG_ETH_SPI_ETHERNET_W5500				true

//////////////////////////////////////////////////////////////

#include <WiFi.h>
#include <WebServer.h> // Introduce corresponding libraries

#include <hal/spi_types.h>

//////////////////////////////////////////////////////////////

#if USING_ESP32_S3

	#if !defined(ETH_SPI_HOST)
		// Using SPI2 for ESP32_SC, but name is SPI3_HOST
		#define ETH_SPI_HOST            SPI3_HOST
	#endif

	#if !defined(SPI_CLOCK_MHZ)
		// Using 25MHz for W5500, 14MHz for W5100
		#define SPI_CLOCK_MHZ       25
	#endif

	#if !defined(INT_GPIO)
		#define INT_GPIO            4
	#endif

	#if !defined(MISO_GPIO)
		#define MISO_GPIO           13
	#endif

	#if !defined(MOSI_GPIO)
		#define MOSI_GPIO           11
	#endif

	#if !defined(SCK_GPIO)
		#define SCK_GPIO            12
	#endif

	#if !defined(CS_GPIO)
		#define CS_GPIO             10
	#endif

//////////////////////////////////////////////////////////////

#elif USING_ESP32_S2

	#if !defined(ETH_SPI_HOST)
		// Using SPI1 for ESP32_SC, but name is SPI2_HOST
		#define ETH_SPI_HOST            SPI3_HOST
	#endif

	#if !defined(SPI_CLOCK_MHZ)
		// Using 25MHz for W5500, 14MHz for W5100
		#define SPI_CLOCK_MHZ       25
	#endif

	#if !defined(INT_GPIO)
		#define INT_GPIO            4
	#endif

	#if !defined(MISO_GPIO)
		#define MISO_GPIO           37
	#endif

	#if !defined(MOSI_GPIO)
		#define MOSI_GPIO           35
	#endif

	#if !defined(SCK_GPIO)
		#define SCK_GPIO            36
	#endif

	#if !defined(CS_GPIO)
		#define CS_GPIO             34
	#endif

//////////////////////////////////////////////////////////////

#elif USING_ESP32_C3

	#if !defined(ETH_SPI_HOST)
		// Using SPI2 for ESP32_SC, but name is SPI3_HOST
		#define ETH_SPI_HOST            SPI2_HOST
	#endif

	#if !defined(SPI_CLOCK_MHZ)
		// Using 25MHz for W5500, 14MHz for W5100
		#define SPI_CLOCK_MHZ       25
	#endif

	#if !defined(INT_GPIO)
		#define INT_GPIO            10
	#endif

	#if !defined(MISO_GPIO)
		#define MISO_GPIO           5
	#endif

	#if !defined(MOSI_GPIO)
		#define MOSI_GPIO           6
	#endif

	#if !defined(SCK_GPIO)
		#define SCK_GPIO        		4    
	#endif

	#if !defined(CS_GPIO)
		#define CS_GPIO             7
	#endif
	
#endif

//////////////////////////////////////////////////////////////

extern bool ESP32_W5500_eth_connected;

extern void ESP32_W5500_onEvent();

extern void ESP32_W5500_waitForConnect();

extern bool ESP32_W5500_isConnected();

extern void ESP32_W5500_event(WiFiEvent_t event);

//////////////////////////////////////////////////////////////

#endif    // WEBSERVER_ESP32_SC_W5500_HPP
